### PR TITLE
Introduce QuarkusProject and Quarkus project zipping in codegen

### PR DESCRIPTION
**Pull request by ia3andy**
This is part of #8178

This is a good one :-), things starts to get nicer and cleaner (but not quite there yet ^^)..

Introduce `QuarkusProject` in codegen to reduce ProjectWriter & BuildTool coupling, it will slowly replace `BuildFile` and `ProjectWriter`.

I also uniformised the quarkus-cli (I am not sure if it's used) by giving the project folder instead of the pom.xml, it should now also work with Gradle.

Deleted `ZipProjectWriter` which is not used and will be replaced by a way to zip a Quarkus project folder.

### CODE REVIEW

**Comment by ia3andy** [comment]
Putting in WIP, I am going to introduce the new folder zipping as another commit.

File: devtools/maven/src/main/java/io/quarkus/maven/CreateProjectMojo.java [hunk 1]
```diff
@@ -236,10 +235,9 @@ public void execute() throws MojoExecutionException {
         }
     }
 
-    private void createGradleWrapper(QuarkusPlatformDescriptor platform,
-            ProjectWriter writer) {
+    private void createGradleWrapper(QuarkusPlatformDescriptor platform, Path projectFolderPath) {
         try {
-            writer.mkdirs("gradle/wrapper");
+            Files.createDirectories(projectFolderPath.resolve("gradle/wrapper"));
```
Comment (ia3andy) [inline devtools/maven/src/main/java/io/quarkus/maven/CreateProjectMojo.java:240]: I need to check that it still works..

**Comment by ia3andy** [comment]
@aloubyansky there is a bug with the maven plugin due to `ProjectWriter.close()`, I'll figure it out tomorrow..

**Comment by ia3andy** [comment]
ok, found it, it's nothing, I'll push tomorrow..

File: devtools/maven/src/main/java/io/quarkus/maven/CreateProjectMojo.java (hunk 1 above)
Comment (ia3andy) [reply devtools/maven/src/main/java/io/quarkus/maven/CreateProjectMojo.java:240]: Working..

File: independent-projects/tools/common/src/main/java/io/quarkus/cli/commands/CreateProjectCommandHandler.java [hunk 2]
```diff
@@ -20,10 +21,8 @@
 
     @Override
     public QuarkusCommandOutcome execute(QuarkusCommandInvocation invocation) throws QuarkusCommandException {
-        final ProjectWriter projectWriter = invocation.getProjectWriter();
-        if (projectWriter == null) {
-            throw new IllegalStateException("Project writer has not been provided");
-        }
+        final ProjectWriter projectWriter = new FileProjectWriter(
+                invocation.getQuarkusProject().getProjectFolderPath().toFile());
```
Comment (aloubyansky) [inline independent-projects/tools/common/src/main/java/io/quarkus/cli/commands/CreateProjectCommandHandler.java:25]: Just a note, previously a single invocation could have been passed to multiple handlers (e.g. create followed by add extensions) that would use the same project writer instance. It doesn't mean it has to stay that way.

**Comment by aloubyansky** [comment]
One thing to keep in mind is the public API should be backward compatible. Our plugins are looking for the latest platform in a given version range. So, inside that range the API should stay compatible.
The issue you ran into with 1.4.2.Final is an example of what shouldn't happen to a user.
I am not going to insist on keeping the current API now. But you should keep this point in mind for the future.
BTW, the current version range for the platform is up to 2.0. We did consider changing it to a micro. E.g. [1.5.0, 1.6.a)

**Comment by aloubyansky** [comment]
The PR is marked as a WIP, so I am not approving it yet.

File: independent-projects/tools/common/src/main/java/io/quarkus/cli/commands/CreateProjectCommandHandler.java (hunk 2 above)
Comment (ia3andy) [reply independent-projects/tools/common/src/main/java/io/quarkus/cli/commands/CreateProjectCommandHandler.java:25]: I am not there yet, but to make it more consistent and simple, my intent is to go toward atomic operations (without having to deal with the writer close/save from outside):
- create (directly with or without extensions)
- add extension(s)
- remove extension(s)

Using the same writer would make it more performant when doing multiple of those, but I guess there is no use case for it.

One good example of the risk of having this close/save handled externally is that (as I did without knowing) having multiple writer and closing the one without the modification after ending up with reseting the modification 😅

WDYT?

**Comment by ia3andy** [comment]
> One thing to keep in mind is the public API should be backward compatible. Our plugins are looking for the latest platform in a given version range. So, inside that range the API should stay compatible.
> The issue you ran into with 1.4.2.Final is an example of what shouldn't happen to a user.
> I am not going to insist on keeping the current API now. But you should keep this point in mind for the future.
> BTW, the current version range for the platform is up to 2.0. We did consider changing it to a micro. E.g. [1.5.0, 1.6.a)

Yeah, I was wondering... It would be good to be allowed to break it in one shot for this refactor? I am not sure if this part of the code is used externally by other than code.quarkus.io? 

Once it's is clean, we can of course lock it..

**Comment by ia3andy** [comment]
Ok I added the project zip compression utils https://github.com/quarkusio/quarkus/pull/9631/commits/f44f9801fdda4a64e87536011270660ca0242dff (inspired from code.quarkus.io with exec support for wrapper files)

And I fixed the BuildFile closing bug by removing the close in `BuildFileMojoBase`, the close should be done by the `doExecute()`

File: independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/QuarkusProject.java [hunk 3]
```diff
@@ -33,12 +33,9 @@ public static QuarkusProject of(final Path projectFolderPath, final QuarkusPlatf
 
     public static QuarkusProject resolveExistingProject(final Path projectFolderPath,
             final QuarkusPlatformDescriptor descriptor) {
-        if (projectFolderPath.resolve("pom.xml").toFile().exists()) {
-            return of(projectFolderPath, descriptor, BuildTool.MAVEN);
-        } else if (projectFolderPath.resolve("build.gradle").toFile().exists()
-                || projectFolderPath.resolve("build.gradle.kts").toFile().exists()) {
-            return of(projectFolderPath, descriptor, BuildTool.GRADLE);
-        }
+        QuarkusProject x = of(projectFolderPath, descriptor, resolveExistingProjectBuildTool(projectFolderPath));
```
Comment (ia3andy) [inline independent-projects/tools/common/src/main/java/io/quarkus/devtools/project/QuarkusProject.java:36]: Oups this is a automatic refactor mistake

**Comment by ia3andy** [comment]
@aloubyansky CI is failing because quay seems to be down..

**Review by aloubyansky — APPROVED** [review]

**Comment by maxandersen** [comment]
> Yeah, I was wondering... It would be good to be allowed to break it in one shot for this refactor? I am not sure if this part of the code is used externally by other than code.quarkus.io?

this is something we should discuss the timing off. I fully agree it needs to happen - but want to try and make it happen in just one release if at all possible.

**Comment by maxandersen** [comment]
> BTW, the current version range for the platform is up to 2.0. We did consider changing it to a micro. E.g. [1.5.0, 1.6.a)

shouldn't we just do this ? it would be `[1.5.0,1.6.0)`

other only alternative I can think of is the introduction of a separate file where we externalize the range we check so we can adjust it after release and be able to update it once we know incompatibility happens.

**Pull request merged**